pkg/controller/bookkeepercluster/upgrade_test.go

```diff
@@ -0,0 +1,341 @@
+					foundBookeeper.Status.TargetVersion = foundBookeeper.Spec.Version
+					foundBookeeper.Status.CurrentVersion = "0.5.1"
+					foundBookeeper.Status.Init()
+					foundBookeeper.Status.SetUpgradingConditionTrue(" ", " ")
+					r.client.Update(context.TODO(), foundBookeeper)
+					err = r.syncClusterVersion(foundBookeeper)
+				})
+				It("Error should be nil", func() {
+					Ω(err).Should(BeNil())
+				})
+			})
+			Context("rollbackClusterVersion", func() {
+				var (
+					err            error
+					foundBookeeper *v1alpha1.BookkeeperCluster
+				)
+				BeforeEach(func() {
+					foundBookeeper = &v1alpha1.BookkeeperCluster{}
+					_ = client.Get(context.TODO(), req.NamespacedName, foundBookeeper)
+					foundBookeeper.Status.UpdateProgress("UpgradeErrorReason", "")
+					r.client.Update(context.TODO(), foundBookeeper)
+					err = r.rollbackClusterVersion(foundBookeeper, "0.6.1")
+				})
+				It("Error should not be nil", func() {
+					Ω(strings.ContainsAny(err.Error(), "failed to get statefulset ()")).Should(Equal(true))
+				})
+			})
+			Context("rollbackClusterVersion syncCompleted return true ", func() {
+				var (
+					err            error
+					foundBookeeper *v1alpha1.BookkeeperCluster
+				)
+				BeforeEach(func() {
+					_, _ = r.Reconcile(req)
+					foundBookeeper = &v1alpha1.BookkeeperCluster{}
+					_ = client.Get(context.TODO(), req.NamespacedName, foundBookeeper)
+					foundBookeeper.Status.TargetVersion = foundBookeeper.Spec.Version
+					foundBookeeper.Status.Init()
+					foundBookeeper.Status.UpdateProgress("UpgradeErrorReason", "")
+					r.client.Update(context.TODO(), foundBookeeper)
+					err = r.rollbackClusterVersion(foundBookeeper, "0.6.1")
+				})
+				It("Error should be nil", func() {
+					Ω(err).Should(BeNil())
+				})
+			})
+			Context("checkUpdatedPods", func() {
+				var (
+					boolean bool
+					pod     []*corev1.Pod
+				)
+				BeforeEach(func() {
+					boolean, err = r.checkUpdatedPods(pod, "0.7.1")
+				})
+				It("Error should be nil and bool value should be true", func() {
+					Ω(err).Should(BeNil())
+					Ω(boolean).Should(Equal(true))
+				})
+			})
+			Context("getOneOutdatedPod", func() {
+				var sts *appsv1.StatefulSet
+				BeforeEach(func() {
+					sts = &appsv1.StatefulSet{}
+					r.client.Get(context.TODO(), types.NamespacedName{Name: util.StatefulSetNameForBookie(b.Name), Namespace: b.Namespace}, sts)
+					_, err = r.getOneOutdatedPod(sts, "0.6.1")
+				})
+				It("Error should be nil", func() {
+					Ω(err).Should(BeNil())
+				})
+			})
+			Context("checkUpdatedPods with faulty pod", func() {
+				var (
+					boolean bool
+					pod     []*corev1.Pod
+				)
+				BeforeEach(func() {
+					testpod := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: "default", Name: "test"}, Spec: v1.PodSpec{Containers: []v1.Container{{Image: "testimage"}}},
+						Status: v1.PodStatus{
+							ContainerStatuses: []v1.ContainerStatus{
+								{
+									Name: "test",
+									State: v1.ContainerState{
```

Comment:
you can check for pod non-faulty case as well here

Reply:
Done.